### PR TITLE
docs: git best practices

### PR DESCRIPTION
**Pull request by gforcada**
Documentation regarding some tips and tricks on how to do better
development with git, be more respectful for the others that will
review your code and giving some advices on how to make so.

Fixes: https://github.com/plone/buildout.coredev/issues/67

### CODE REVIEW

**Comment by jean** [comment]
@gforcada thanks for this.

File: docs/git.rst [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+git best practices
+==================
+Collection of best practices that would help you and the overall community to work with git more efficiently.
+
+General guidelines
+==================
+
+Pulling code
+------------
+*ALWAYS* (in capital letters) do a :command:`git pull -r` when fetching new code,
```
Comment (do3cc) [inline docs/git.rst]: Here I wonder what -r does, but instead of explaining that, you go on with how to automate that thing

File: docs/git.rst [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+--------------
+For commit messages see:
+`plone API guidelines<http://docs.plone.org/external/plone.api/docs/contribute/conventions.html#git-commit-message-style>`_.
+
+Adding references to issues
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Always add the full URL to the issue/pull request you are fixing/referring to.
+
+Maybe within the git repository it makes sense,
+but as soon as you are outside of it, it will not.
```
Comment (do3cc) [inline docs/git.rst]: I guess you rephrased something and now this sentence does not make sense right now

**Comment by do3cc** [comment]
I am not a huge fan of doctoring commits after the fact. Mostly because I find it very hard to do so and I am not very proficient with it. Some things I miss: 
When I want to do clean commits I must prevent myself from just starting to modify things I notice.
I regulary start to work on something and during that work I notice I have to fix other things too. Here is an example: https://github.com/collective/collective.solr/pull/40 I would not know how to solve this.
I like the idea of changing a line only in one commit within a PR. What tools are there to help me check for this?

Finally, I am a big +1 on this, but I'd hope to receive more help on how to do this.

File: docs/git.rst (hunk 1 above)
Comment (gforcada) [inline docs/git.rst]: Good point, will go over it again. `-r` is to do a rebase after fetching the code from the repository.

So basically `git pull` in a vanilla git install actually does:

```
git fetch
git merge
```

With the `-r` switch you are actually doing:

```
git fetch
git rebase origin/XXX # wherever the branch you happen to be.
```

File: docs/git.rst (hunk 2 above)
Comment (gforcada) [inline docs/git.rst]: Most probably, as I said I wanted to get a first version out and later review it.

In this section I was referring to the fact that if your commit message says:

```
Fix issue #74
```

You can relate that issue 74 to the current project where the commit was pushed, but alas, p.a.discussion (for the sake of an example) does not have issues enabled on github, so where the hell is this issue 74 from?

My suggestion is to instead write it like:

```
Fix issue about encoding

La la la la la la la
la la la la li li li li li

Fixes issue: https://github.com/plone/Products.CMFPlone/issue/74
```

**Comment by gforcada** [comment]
@do3cc  thanks for the comments and review!

About doctoring commits, I kind of understand, but still one year later, if you go stare at the code and want to understand what was going on there you will most probably end up looking at `git blame` and looking at those commits, specially for the commit messages and related files (maybe for the tests that show how the logic is supposed to work).

So at the end of the day (or year) the more effort you put on properly explaining your commits and the more curated they are, the easier it will be for other to look at them, understand what was going on and so on.

Where I work (der Freitag) we use viewlets _A LOT_, so as I'm really lazy to remember everything, I just need to remember the last time I created one and literally copy&paste&adapt the commits from the last viewlet I created because everything is there: tests, code, ZCML registration, GS profile, upgrade step, and so on.

About how to do clean commits: whenever I found something to be cleaned up I do:
- `git stash` (to save my current changes temporally)
- change back to master (if I'm working on a branch, which usually you should :))
- do the cleanups && push those changes
- get back to the branch
- rebase on top of master (`git rebase master`)
- `git stash pop` (get back the temporal changes)
- _if needed_, fix the conflicts on the code that was stashed and the cleanups that you did on master

Seems a lot of steps but is quite fast to do so.

I have another section pending to be added also about manually reverting pull requests if github doesn't allow you to do so. So I will add this cleanup strategy also.

**Comment by do3cc** [comment]
He @gforcada thanks for the description. Worth a try and imho it would be good if this is part of the docs. 
In this particular PR I mentioned this wouldn't have been possible, but rewriting all tests is kind of big after all.

File: docs/git.rst (hunk 1 above)
Comment (adamcheasley) [inline docs/git.rst]: I agree with this point, but as @do3cc says, you should possibly mention what the argument does. Even just mentioning the word 'rebase' might be enough.

Comment (adamcheasley) [inline docs/git.rst]: Sorry, ignore that last comment. I see that you have a link to further explanation. This is probably enough.

File: docs/git.rst [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+or you say it like "it does XXX and YYY",
+you most probably need to split that commit into,
+at least,
+two or three commits.
+
+That doesn't mean that a +20 files or +100 lines of code changes are bad per se,
+you may be doing a simple refactoring across lots of files,
+that's fine and good actually,
+as long as that commit is just and only about that refactoring,
+and not a refactoring plus moving things around plus fixing some bugs plus adding some docs plus a new cool feature plus fixing typos on documentation.
```
Comment (adamcheasley) [inline docs/git.rst]: Ironically, this sentence is quite long and could probably be broken down into two shorter sentences. It's a little hard to follow at present.

**Comment by adamcheasley** [comment]
Thanks for this @gforcada. I think this document is something that is desperately needed within the community.

**Comment by pilz** [comment]
I am very happy that you did this! I strongly believe that many many people want to follow good practice but it is difficult to know what that is. Your document is a big step! Thank you!

**Comment by wengole** [comment]
@do3cc With regards your issue of wanting changing more things than you should in one commit, you might want to read up on partial commits. UI tools help here tremendously, but it can be done from the command line quite simply too.

For UI, on Linux I use git-cola and I've seen the Mac users here (netsight) use gitx (not to be confused with gitx on Linux which is quite different)

Reference;
http://stackoverflow.com/a/1085191
http://git-scm.com/docs/git-add#_interactive_mode

File: docs/git.rst (hunk 1 above)
Comment (jean) [inline docs/git.rst]: @adamcheasley @gforcada the option can be spelled `--rebase` instead of `-r` ... I think there's seldom any reason to use the short forms in docs.

Comment (gforcada) [inline docs/git.rst]: Good point @jean 

Actually as I got the idea that it was a bit to much into the point, right now I'm writing a little introduction, hopefully it doesn't end up being a duplicate of git's own book: http://git-scm.com/book/en/v2 :)

**Comment by gforcada** [comment]
@do3cc @adamcheasley and others, I just pushed a reworked version of it, hopefully addressing your (really good) comments. If you could give it a go it would be much appreciated.

**Comment by gforcada** [comment]
I see that a proof read is more than needed, but at 2am I don't feel like it, so please, bear with the grammar errors on the meantime, sorry!

**Comment by adamcheasley** [comment]
Thanks @gforcada. This looks really good. Like you say, there are a few grammatical errors, but I held off commenting on those until you've had time for a proof read.

**Comment by gforcada** [comment]
@adamcheasley it would be really appreciated (no native English speaker here as you can see) if you can either fix or just add comments on them in github, thanks in advance!

**Comment by adamcheasley** [comment]
No problem, will do.

File: docs/git.rst [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+tags
+   A name attached to a single commit.
+
+``HEAD``
+   A pointer that always tells you where you are
+   (extremely useful when doing some operations).
+
+The index
+  A temporal staging storage with changes on files that are pending to be added to a commit.
+  If your git output is colored,
```
Comment (adamcheasley) [inline docs/git.rst:60]: "coloured"! "colored" is the US spelling.

File: docs/git.rst [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+Stash
+  Temporal storage for changes,
+  again,
+  extremely useful in some scenarios,
+  see further below for examples.
+
+Branches
+--------
+Another great feature of DVCS is cheap branching,
+i.e. make it effortlessly and really useful to use branches.
```
Comment (adamcheasley) [inline docs/git.rst]: I think this would read better as:
"i.e. branching in git is effortless and really useful."

File: docs/git.rst (hunk 4 above)
Comment (do3cc) [inline docs/git.rst:60]: Do we have enough us speakers to make readable for them too? SCNR ;-)

File: docs/git.rst [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+No more working on the master branch always as is too much effort to work on branches.
```
Comment (adamcheasley) [inline docs/git.rst]: Possibly consider changing to:
"As it's no longer too much effort to branch, there is no need to always work on the master branch. A developer can branch easily for each fix/feature."

File: docs/git.rst [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+
+Branches allows you to tinker with your changes while keeping the master branch clean.
```
Comment (adamcheasley) [inline docs/git.rst]: "Branches allow you". There is an erroneous 's' on the end of allow there.

File: docs/git.rst [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+
+Not only that,
+it also allows to keep modifying your changes until you and your peers are fine with them.
```
Comment (adamcheasley) [inline docs/git.rst]: "it also allows you to keep..."

File: docs/git.rst [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+      As long as you don't remove the ``.git`` folder,
+      even if you remove the file you just added,
+      the changes you made before doing ``git add`` are still there ready to be recovered at any time!
+
+status
+   Get an overview of the repository status.
+
+   If there are files on the index,
+   or files not tracked by git,
+   or the status of your branch in regards of the remote one,
```
Comment (adamcheasley) [inline docs/git.rst]: "or the status of your local branch with regards to the remote"

File: docs/git.rst [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+   An image is worth 1000 words: http://git-scm.com/docs/git-rebase
+
+checkout
+   Change to the given branch or get the given file to its latest committed version.
+
+   .. note::
+      If git is criticized for being complex,
+      this command is one of the main sources of complains.
+
+      It is so *powerful* yet apparently so *random* that makes people crazy.
```
Comment (adamcheasley) [inline docs/git.rst]: I'm not entirely clear on this sentence. There's nothing really random about this command. It might be helpful to compare this command to svn's 'switch'.

File: docs/git.rst [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+How to solve it?
+^^^^^^^^^^^^^^^^
+*ALWAYS* do a :command:`git pull --rebase` when fetching new code,
+configure git to do always so with::
+
+    git config branch.autosetuprebase always # add the --global switch to make it default everywhere
+
+This way you do not introduce new extra commits and the git history is kept as simple as possible.
+
+This is specially important when trying to understand why some changes were made,
```
Comment (adamcheasley) [inline docs/git.rst]: "This is especially important..."

File: docs/git.rst (hunk 4 above)
Comment (davisagli) [inline docs/git.rst:60]: As a developer on international open source projects I speak fluent butchered English. :-p

File: docs/git.rst [hunk 12]
```diff
@@ -0,0 +1,611 @@
...
+
+This is the time when you spot typos,
+pep8 errors,
+misaligned code,
+lack of docstrings in methods,
+that a permission is not defined on Generic Setup,
+that an upgrade should be needed...
+
+Remember that the first code review is the one you do on your own.
+Some inspiration/better phrased:
```
Comment (adamcheasley) [inline docs/git.rst]: should be "better phrasing"

File: docs/git.rst [hunk 13]
```diff
@@ -0,0 +1,611 @@
...
+Something that should mostly be a verification of a checklist like:
+
+- the browser view is registered on ZCML?
+- is there an interface for that form?
+- the pt and py are there?
+- ...
+
+Turns instead into a list of questions:
+
+- why this interface is renamed here if has nothing to do with this adapter?
```
Comment (adamcheasley) [inline docs/git.rst]: "Why is this interface renamed here if it has nothing to do with this adapter?"

File: docs/git.rst [hunk 14]
```diff
@@ -0,0 +1,611 @@
...
+
+- refactoring code
+- moving things around
+- fixing some bugs while at it
+- adding some docs
+- a new cool feature
+- fixing typos on documentation
+- pep8 fixes
+
+Is completely fine to make refactorings.
```
Comment (adamcheasley) [inline docs/git.rst]: "It is absolutely fine to refactor."

File: docs/git.rst [hunk 15]
```diff
@@ -0,0 +1,611 @@
...
+
+And this is actually to help both your present self and your +5 years from now that will have to refactor that code of yours,
+and maybe is struggling to understand what was going on there.
+
+Following this advice will:
+
+- keep things simple where there's no gain on adding complexity
```
Comment (adamcheasley) [inline docs/git.rst]: _in_ adding complexity...

File: docs/git.rst [hunk 16]
```diff
@@ -0,0 +1,611 @@
...
+- make your changes easy to be reviewed
+- make later on lookups on those changes easy to follow
```
Comment (adamcheasley) [inline docs/git.rst:380]: Could you clarify this last point for me?

File: docs/git.rst [hunk 17]
```diff
@@ -0,0 +1,611 @@
...
+Question:
+if it's BBB only,
+where is the new place to look for that interface now?
+
+The problem is that,
+in this case Martin,
+wrote that in 2009,
+so most probably once a refactor of that package is done later on 2015,
+Martin is no longer around,
+and if he would be,
```
Comment (adamcheasley) [inline docs/git.rst]: "and if he was,"

File: docs/git.rst [hunk 18]
```diff
@@ -0,0 +1,611 @@
...
+
+Ask yourself a question:
+if someone comes to you asking details about a random commit done by you +5 years ago,
+what will you reply?
+
+Try that,
+get one project that you worked 5 years ago,
+get a random commit and:
+
+- see if by just reading the commit message already gives enough information of what changes are in
```
Comment (adamcheasley) [inline docs/git.rst]: "See if, just by reading the commit message, you are given enough information of what changes have been made, when comparing the commit message and the actual code. Does the commit message match the code changed?"

File: docs/git.rst [hunk 19]
```diff
@@ -0,0 +1,611 @@
...
+if you do merges with master,
+you are actually spreading your pull request into more commits,
+and at the end making it more difficult to track what was changed.
+
+On top of that,
+the commit history is more complex to follow.
+
+See the history example above: :ref:`general-guidelines-label`.
+
+Unfortunately the flat view from GitHub prevents to see that,
```
Comment (adamcheasley) [inline docs/git.rst]: "Unfortunately the flat view from GitHub prevents us from seeing that"

File: docs/git.rst [hunk 20]
```diff
@@ -0,0 +1,611 @@
...
+is to make him/her spend time reviewing some code changes that one the next commit are changed again to do something else.
+
+It will not only make your commits smaller,
+but it will also make it easy to do atomic commits.
+
+No cleanup commits please
+-------------------------
+*On the context of a pull request*
+
+Ask yourself: which relation has a cleanup commit,
```
Comment (adamcheasley) [inline docs/git.rst]: "Ask yourself: What relation does a cleanup commit, say pep8 fixes or other code analysis fixes, have with your pull request?"

File: docs/git.rst [hunk 21]
```diff
@@ -0,0 +1,611 @@
...
+    * 11ba28c Last fix, finally
+    * 11ba28c Fix tests, again
+    * 11ba28c Fix tests
+    * 11ba28c Do something fancy
+    * 11ba28c Failing test, we are doing TDD right?
+
+Only tells you that the author did not take care at all about the one who will review it,
+and specially about the person that in +5 years will try to understand that test.
+Specially because now the test is not only spread between 4 commits,
+but most probably during those 5 years it already had some refactorings,
```
Comment (adamcheasley) [inline docs/git.rst]: "but most probably during those 5 years it has already been refactored,"

**Comment by adamcheasley** [comment]
Done! I've just realised, I could have helped you out a bit there and just made the changes in the branch. If you'd like me to do that, just let me know.

File: docs/git.rst (hunk 16 above)
Comment (gforcada) [inline docs/git.rst:380]: Sorry for the late answer: I'm trying to say that if "one commit does one thing" if you ever have to do something similar (say create a viewlet) it will make your life easier as it will be a matter of copy/pasting that previous commit that you can look it up.

Does that make sense?

**Comment by gforcada** [comment]
@adamcheasley I just pushed your fixes, many thanks!

If no one objects I think it can be merged.

**Comment by do3cc** [comment]
heh, I was too slow.
I finished the reading and I am happy with the recipes extension.
What I am not sure is if the doc followed the guidelines, especially this one:
http://docs.plone.org/about/styleguide.html#line-length-translations

At first I thought you tried to stay within 80 chars but then I saw that this is not the case. Right now some phrases are written in so many tiny lines that it is hard to read in source. Like here: https://github.com/plone/buildout.coredev/pull/71/files#diff-5a9477c2b3ec283ab80fd30c2fa4726dR69
